extensions/3d.sh

```diff
@@ -0,0 +1,92 @@
+#
+# SPDX-License-Identifier: GPL-2.0
+# Armbian build framework extension
+#
+# Enables 3D and multimedia acceleration for Debian and Ubuntu
+# - 
+#
+
+function extension_prepare_config__3d() {
+
+	[[ "${BUILDING_IMAGE}" != "yes" ]] && return 0
+	EXTRA_IMAGE_SUFFIXES+=("-3d") # Add to the image suffix. # global array
+
+}
+
+function post_install_kernel_debs__3d() {
+
+	declare -a pkgs=("mesa-utils" "mesa-utils-extra" "libglx-mesa0" "libgl1-mesa-dri" "glmark2" "glmark2-wayland" "glmark2-x11" "glmark2-es2-wayland" "glmark2-es2" "glmark2-es2-x11")
+
+	# Do not install those packages on CLI images
+	[[ "${BUILD_DESKTOP}" != "yes" ]] && return 0
+
+	# Old releases whcih are not supported but are still in the system
+	[[ "${RELEASE}" =~ ^(bullseye|buster|focal)$ ]] && return 0
+
+	if [[ "${LINUXFAMILY}" =~ ^(rockchip-rk3588|rk35xx)$ && "$BRANCH" =~ ^(legacy)$ && "${RELEASE}" =~ ^(jammy)$ ]]; then
+
+		display_alert "Adding amazingfated's rk3588 PPAs" "${EXTENSION}" "info"
+		do_with_retries 3 chroot_sdcard add-apt-repository ppa:liujianfeng1994/panfork-mesa --yes --no-update
+
+		display_alert "Pinning amazingfated's rk3588 PPAs" "${EXTENSION}" "info"
+		cat <<- EOF > "${SDCARD}"/etc/apt/preferences.d/amazingfated-rk3588-panfork-pin
+		Package: *
+		Pin: release o=LP-PPA-liujianfeng1994-panfork-mesa
+		Pin-Priority: 1001
+		EOF
+
+	elif [[ "${DISTRIBUTION}" == "Ubuntu" ]]; then
+
+		display_alert "Adding oibaf PPAs" "${EXTENSION}" "info"
+		do_with_retries 3 chroot_sdcard add-apt-repository ppa:oibaf/graphics-drivers --yes --no-update
+
+		display_alert "Pinning oibaf PPAs" "${EXTENSION}" "info"
+		cat <<- EOF > "${SDCARD}"/etc/apt/preferences.d/mesa-oibaf-graphics-drivers-pin
+		Package: *
+		Pin: release o=LP-PPA-oibaf-graphics-drivers
+		Pin-Priority: 1001
+		EOF
+
+	fi
+
+	# This should work on all distributions where mesa
+	[[ "${LINUXFAMILY}" == "rockchip-rk3588" && "${LINUXFAMILY}" == "rk35xx" && "$BRANCH" == vendor ]] && declare -g DEFAULT_OVERLAYS="panthor-gpu"
+
+	if [[ "${LINUXFAMILY}" =~ ^(rockchip-rk3588|rk35xx|rockchip64)$ && "${RELEASE}" =~ ^(jammy|noble)$ ]]; then
```

Comment:
We should limit rockchip-multimedia ppa to legacy and vendor branch.

Reply:
> We should limit rockchip-multimedia ppa to legacy and vendor branch.

So mainline based rockchip is pointless to have this?

Reply:
Yes. This ppa is made for userspace mpp apps, and only rockchip's kernel has the kernel driver.